php-frontend/src/main/java/org/sonar/php/parser/TreeFactory.java

```diff
@@ -0,0 +1,1702 @@
+    InternalSyntaxToken openCurlyBrace,
+    Optional<List<TraitAdaptationStatementTree>> adaptations,
+    InternalSyntaxToken closeCurlyBrace
+  ) {
+    return new TraitUseStatementTreeImpl(useToken, traits, openCurlyBrace, optionalList(adaptations), closeCurlyBrace);
+  }
+
+  public TraitPrecedenceTree traitPrecedence(
+    TraitMethodReferenceTree methodReference,
+    InternalSyntaxToken insteadOfToken,
+    SeparatedList<NamespaceNameTree> traits,
+    InternalSyntaxToken eosToken
+  ) {
+    return new TraitPrecedenceTreeImpl(methodReference, insteadOfToken, traits, eosToken);
+  }
+
+  public TraitAliasTree traitAlias(
+    TraitMethodReferenceTree methodReference,
+    InternalSyntaxToken asToken,
+    Optional<SyntaxToken> modifier,
+    IdentifierTree alias,
+    InternalSyntaxToken eos
+  ) {
+    return new TraitAliasTreeImpl(methodReference, asToken, modifier.orNull(), alias, eos);
+  }
+
+  public TraitAliasTree traitAlias(
+    TraitMethodReferenceTree methodReference,
+    InternalSyntaxToken asToken,
+    SyntaxToken modifier,
+    InternalSyntaxToken eos
+  ) {
+    return new TraitAliasTreeImpl(methodReference, asToken, modifier, null, eos);
+  }
+
+  public TraitMethodReferenceTree traitMethodReference(InternalSyntaxToken identifier) {
+    return new TraitMethodReferenceTreeImpl(new IdentifierTreeImpl(identifier));
+  }
+
+  public TraitMethodReferenceTree traitMethodReference(NamespaceNameTree trait, InternalSyntaxToken doubleColonToken, InternalSyntaxToken identifier) {
+    return new TraitMethodReferenceTreeImpl(trait, doubleColonToken, new IdentifierTreeImpl(identifier));
+  }
+
+  public ClassDeclarationTree interfaceDeclaration(
+    InternalSyntaxToken interfaceToken, IdentifierTree name,
+    Optional<Tuple<InternalSyntaxToken, SeparatedList<NamespaceNameTree>>> extendsClause,
+    InternalSyntaxToken openCurlyBraceToken, Optional<List<ClassMemberTree>> members, InternalSyntaxToken closeCurlyBraceToken
+  ) {
+    InternalSyntaxToken extendsToken = null;
+    SeparatedList<NamespaceNameTree> interfaceList = SeparatedList.empty();
+    if (extendsClause.isPresent()) {
+      extendsToken = extendsClause.get().first();
+      interfaceList = extendsClause.get().second();
+    }
+    return ClassDeclarationTreeImpl.createInterface(
+      interfaceToken,
+      name,
+      extendsToken,
+      interfaceList,
+      openCurlyBraceToken,
+      optionalList(members),
+      closeCurlyBraceToken
+    );
+  }
+
+  public ClassDeclarationTree traitDeclaration(
+    InternalSyntaxToken traitToken, IdentifierTree name,
+    InternalSyntaxToken openCurlyBraceToken, Optional<List<ClassMemberTree>> members, InternalSyntaxToken closeCurlyBraceToken
+  ) {
+    return ClassDeclarationTreeImpl.createTrait(
+      traitToken,
+      name,
+      openCurlyBraceToken,
+      optionalList(members),
+      closeCurlyBraceToken
+    );
+  }
+
+  public ClassDeclarationTree classDeclaration(
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Method has 8 parameters, which is greater than 7 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00107)